### PR TITLE
Fixes #49

### LINKED ISSUE #49: Run Single Test runs all tests in a file
Hello,

I noticed that when I try to run a single test method with cursor on it, it runs all tests in the current file.

The problem seems to be affecting only when method names don't start with `test_`
```
   /** @test **/
    public function get_users_first_name_from_name()
    {
    }
```

This works as expected..
```
    public function test_get_users_first_name_from_name()
    {
    }
```

### PR DESCRIPTION
**Pull request by pryley**
Fixes #49 

- Returns an array of selected test method names.
- Selection can be anywhere inside one or more test methods.
- If no selection is found inside any test method, then all test method names are returned.

I'm not too familiar with Python, but this seems to do the trick nicely.

### CODE REVIEW

**Comment by pryley** [comment]
Here are visual examples of the changes in this pull-request for the "PHPUnit: Run Single Test" command (Cmd+Shift+R on Mac): 

**1. Run the test of the method where the cursor is positioned in.**

image

**2. Run the tests of any methods where the cursor is positioned (multi-selection).**

image

**3. Run all tests in the file if the cursor is not positioned in any method.**

image

File: plugin.py [hunk 1]
```diff
@@ -404,18 +404,29 @@ def run(self):
...
+        for region in view.sel():
+            for i, area in enumerate(function_areas):
+                if not area.a <= region.a <= area.b:
+                    continue
+                word = view.substr(function_regions[i])
+                if is_valid_php_identifier(word):
+                    method_names.append(word)
+                break
+
+        if not method_names:
```
Comment (gerardroche) [inline plugin.py]: If no selection then return and empty list `[]` because there's no need to run the test with a filter if none is needed.

File: plugin.py [hunk 2]
```diff
@@ -404,18 +404,29 @@ def run(self):
...
         """
 
-        # @todo should be a scoped selection i.e. is the selection a source.php entity.name.function
         method_names = []
-        for region in view.sel():
-            word = view.substr(view.word(region))
-            if not is_valid_php_identifier(word) or word[:4] != 'test':
-                return None
+        function_areas = view.find_by_selector('meta.function')
+        function_regions = view.find_by_selector('entity.name.function')
```
Comment (gerardroche) [inline plugin.py]: Ensure that areas count is equal regions count.

File: plugin.py [hunk 3]
```diff
@@ -404,18 +404,29 @@ def run(self):
...
```
Comment (gerardroche) [inline plugin.py]: If no function areas or regions are found default to the previous implementation because some syntaxes or old syntaxes like my php-grammar plugin don't have those scopes so this feature will fail on those syntaxes.

File: plugin.py (hunk 2 above)
Comment (pryley) [inline plugin.py]: I implemented this method instead of the previous in order to support unit-tests defined by comments instead of the method prefixed with "test".

File: plugin.py (hunk 3 above)
Comment (pryley) [inline plugin.py]: Agree

File: plugin.py (hunk 1 above)
Comment (pryley) [inline plugin.py]: I opted to run all tests in the file in the selection was none or invalid in order to give the option to run all unit-tests in a single file. "Run all tests" runs all test in the project. There is no command to run all tests in the active file.

Comment (pryley) [inline plugin.py]: image

VS

image

Comment (gerardroche) [inline plugin.py]: Does the first one have 4 test cases in the file?

Currently, running a single test will either:

* run the test case i.e. all tests within the current testcase (usually the convention is to have one test case per file, at least 95% use case)
* runs the test (method under cursor)
* runs the tests (multiple selection test methods)

So returning all the test methods will run the test case with filter (all the tests), which is the same thing as just running the test case with no filter.

File: plugin.py (hunk 2 above)
Comment (gerardroche) [inline plugin.py]: You can probably loop through the areas and instead of looking up the region by area index, find the entity.name.function selector within the area.

**Comment by gerardroche** [comment]
I think this feature is pretty much resolving this feature https://github.com/gerardroche/sublime-phpunit/issues/37

File: plugin.py (hunk 1 above)
Comment (pryley) [inline plugin.py]: No, each file extends a singe TestCase base class; each test case includes multiple tests. It's an early project so there are only a few small tests.

This feature was primary made to resolve #49 so that the plugin would work with _annotated_ tests. By being able to select anywhere inside a test method to run it, rather than specifically the method name, the naming convention of the test method no longer matters for the package to work properly. This is the primary change of this feature. The package works for the developer rather than making the developer modify their coding method to suit the package.

image

The `Run all Tests` command runs _all_ test case files in the project.
The `Run Single Test` runs either the selected test/tests in a test case, or the entire single test case class if the selection is invalid or none.

You are right the new functionality is the same as the old, but it is less restrictive as it tries to suit the _intention_ of the developer. It's much quicker to select anywhere inside of a unit-test method to run it, rather than only selecting inside the method name, especially if selecting more than one test to run.

Comment (gerardroche) [inline plugin.py]: I agree it's much quicker to select anywhere in the method (or nearest #37). 

My point is that a empty list (instead of a list containing al the methods) should be returned if the cursor is not positioned in any method.

Comment (pryley) [inline plugin.py]: Ahh. Thanks for the tip.

File: plugin.py (hunk 2 above)
Comment (pryley) [inline plugin.py]: I am not sure how to find a selector within a region (I'm not a python developer). What about something like this?

```python
method_names = []
function_areas = view.find_by_selector('meta.function')
function_regions = view.find_by_selector('entity.name.function')

for region in view.sel():
    for i, area in enumerate(function_areas):
        if not area.a <= region.a <= area.b:
            continue
        if not i in function_regions and not area.intersects(function_regions[i]):
            continue
        word = view.substr(function_regions[i])
        if is_valid_php_identifier(word):
            method_names.append(word)
        break

return method_names
```

File: plugin.py (hunk 3 above)
Comment (pryley) [inline plugin.py]: Something like this?
```python
for region in view.sel():
    # ...
    if not method_names:
        word = view.substr(view.word(region))
        if is_valid_php_identifier(word) or word[:4] == 'test':
            method_names.append(word)

return method_names
```
